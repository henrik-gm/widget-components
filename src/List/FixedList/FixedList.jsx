import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import styles from './FixedList.scss';
import ItemContainer from '../ItemContainer';

/**
 * Horizontal list component. Items will be stretched or shrunken to fit the container.
 * @memberOf widget-components
 */
class FixedList extends Component {

   /*
    * Constructs.
    * @param {object} props Component properties
    */
   constructor(props) {
      super(props);
      this.state = {
         selected: props.selected
      };
   }

   /*
    * Switches to given tab.
    * @param {number} idx Tab index
    */
   onItemClick(idx) {
      if (this.props.onItemClick) {
         this.props.onItemClick(idx);
      }

      this.setState({ selected: idx });
   }

   /*
    * Renders component.
    * @returns {XML}
    */
   render() {
      return (
         <div className={styles.bar}>
            {Children.map(this.props.children, (child, i) => this.props.renderItemContainer({
               key: i,
               selected: this.state.selected == i,
               onClick: this.onItemClick.bind(this, i),
               children: child
            }))}
         </div>
      );
   }
}

/**
 * @callback FixedList_RenderItemContainer
 * @param {FixedList_RenderItemContainerArgs} args Contains properties which will control the container
 * @returns ReactElement
 *
 * @example <caption>Using custom item container</caption>
 *    ({selected, onClick, onWidth, children}) =>
 *       <CustomItemContainer
 *          selected={selected}
 *          onClick={onClick}
 *          onWidth={onWidth}>
 *             {children}
 *       </CustomItemContainer>
 *
 * @example <caption>Shorthand syntax can be used once function arguments and component properties names match.</caption>
 *    args => <CustomItemContainer {...args}>{args.children}</CustomItemContainer>
 */

/**
 * @name FixedList_RenderItemContainerArgs
 * @property {boolean} selected Controls whether item should be rendered as currently selected or normally
 * @property {function} onClick Called once item has been clicked
 * @property {function} onWidth Called when item width is determined or has been changed
 * @property {ReactElement} children Item contents
 */

/**
 * @property [children] {ReactElement[]} Items list
 * @property [onItemClick] {function(number)} Item click handler. Called with item index argument.
 * @property [selected=0] {number} Initially selected item index
 * @property [renderItemContainer] {FixedList_RenderItemContainer} Function capable of rendering item container. Renders Kambi-styled item container by default.
 */
FixedList.propTypes = {
   children: PropTypes.node,
   onItemClick: PropTypes.func,
   selected: PropTypes.number,
   renderItemContainer: PropTypes.func
};

FixedList.defaultProps = {
   selected: 0,
   renderItemContainer: args =>
      <ItemContainer {...args}>
         {args.children}
      </ItemContainer>
};

export default FixedList;
