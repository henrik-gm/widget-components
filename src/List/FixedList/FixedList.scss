.bar {
   display: flex;

   // distribute equally available width between tab containers
   & > * {
      flex: 1;
   }
}
