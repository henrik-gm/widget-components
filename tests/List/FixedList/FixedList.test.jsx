/* eslint-env jest */
import React from 'react';
import FixedList from '../../../src/List/FixedList/FixedList';
import ReactTestUtils from 'react-addons-test-utils';
import { mount } from 'enzyme';

describe('FixedList view', () => {

   it('renders correctly with default props', () => {
      expect(ReactTestUtils.createRenderer().render(
         <FixedList />
      )).toMatchSnapshot();
   });

   it('renders correctly with one child item', () => {
      expect(ReactTestUtils.createRenderer().render(
         <FixedList>
            <div />
         </FixedList>
      )).toMatchSnapshot();
   });

   it('renders correctly with many child items', () => {
      expect(ReactTestUtils.createRenderer().render(
         <FixedList>
            <div>1</div>
            <div>2</div>
            {[<div key="3">3</div>, <div key="4">4</div>]}
         </FixedList>
      )).toMatchSnapshot();
   });

   it('renders correctly with selected item given arbitrary', () => {
      expect(ReactTestUtils.createRenderer().render(
         <FixedList selected={1}>
            <div>1</div>
            <div>2</div>
         </FixedList>
      )).toMatchSnapshot();
   });

   it('renders correctly custom ItemContainer', () => {
      expect(ReactTestUtils.createRenderer().render(
         <FixedList
            renderItemContainer={({ selected, onClick, onWidth, children }) => <div>{children}</div>}
         >
            <div>1</div>
            <div>2</div>
         </FixedList>
      )).toMatchSnapshot();
   });

});

describe('FixedList interface', () => {

   it('calls onItemClick properly', () => {
      const onItemClickMock = jest.fn();

      const wrapper = mount(
         <FixedList onItemClick={onItemClickMock}>
            <div>1</div>
            <div>2</div>
            <div id="goingToBeClicked">3</div>
         </FixedList>
      );

      expect(onItemClickMock).not.toHaveBeenCalled();

      wrapper.find('#goingToBeClicked').simulate('click');

      expect(onItemClickMock).toHaveBeenCalledTimes(1);
      expect(onItemClickMock).toHaveBeenCalledWith(2);
   });

   it('handles item clicks correctly', () => {
      const renderItemContainerMock = jest.fn(({ selected, onClick, onWidth, children }) => {
         return <div onClick={onClick}>{children}</div>;
      });

      const wrapper = mount(
         <FixedList renderItemContainer={renderItemContainerMock}>
            <div>1</div>
            <div>2</div>
            <div id="goingToBeClicked">3</div>
         </FixedList>
      );

      expect(renderItemContainerMock).toHaveBeenCalledTimes(3);

      renderItemContainerMock.mockClear();

      wrapper.find('#goingToBeClicked').simulate('click');

      expect(renderItemContainerMock).toHaveBeenCalledTimes(3);
   });

});
