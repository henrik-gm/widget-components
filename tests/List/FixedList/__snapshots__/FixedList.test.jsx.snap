// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedList view renders correctly custom ItemContainer 1`] = `
<div
  className="bar"
>
  <div>
    <div>
      1
    </div>
  </div>
  <div>
    <div>
      2
    </div>
  </div>
</div>
`;

exports[`FixedList view renders correctly with default props 1`] = `
<div
  className="bar"
/>
`;

exports[`FixedList view renders correctly with many child items 1`] = `
<div
  className="bar"
>
  <ItemContainer
    onClick={[Function]}
    selected={true}
  >
    <div>
      1
    </div>
  </ItemContainer>
  <ItemContainer
    onClick={[Function]}
    selected={false}
  >
    <div>
      2
    </div>
  </ItemContainer>
  <ItemContainer
    onClick={[Function]}
    selected={false}
  >
    <div>
      3
    </div>
  </ItemContainer>
  <ItemContainer
    onClick={[Function]}
    selected={false}
  >
    <div>
      4
    </div>
  </ItemContainer>
</div>
`;

exports[`FixedList view renders correctly with one child item 1`] = `
<div
  className="bar"
>
  <ItemContainer
    onClick={[Function]}
    selected={true}
  >
    <div />
  </ItemContainer>
</div>
`;

exports[`FixedList view renders correctly with selected item given arbitrary 1`] = `
<div
  className="bar"
>
  <ItemContainer
    onClick={[Function]}
    selected={false}
  >
    <div>
      1
    </div>
  </ItemContainer>
  <ItemContainer
    onClick={[Function]}
    selected={true}
  >
    <div>
      2
    </div>
  </ItemContainer>
</div>
`;
